### PR TITLE
Implementation of a Pagination Support on New Style APIs

### PR DESCRIPTION
**Pull request by LuckyCyborg**
This pull request introduce a new **Pagination API**, in a Laravel-esque style, integrated with the **Database API**, permitting a very simple usage while offering a strong API. 

Example of the **Pagination API** usage:
```php
use DB;

// Get a Paginator instance from the QueryBuilder.
$paginate = DB::table('posts')->paginate(20);

// Add additional query parameters.
$paginate->appends(array(
    'testing' => 1,
    'example' => 'the_example_string',
));

// Get the Pagination Links.
$content = $paginate->links();

// Process the current Page results.
foreach ($paginate->getItems() as $post) {
    $content .= '<h3>' .$post->title .'</h3>';
    $content .= $post->content;
}

echo $content;
```
For a full control on QueryBuilder **paginate()**:
```php
// Get a full customized Paginator instance from the QueryBuilder.
$paginate = DB::table('posts')->paginate(25, array('*')); 
```
Like you see, there is no need to instantiate the **Pagination\Paginator** yourself, the QueryBuilder returning a properly instance, which contains also the expected results for the current Page and accessible with the method **getItems()**, or even using directly the Paginator isntance. For example:
```php
$content = $paginate->links();

// Process the current Page results.
foreach ($paginate as $post) {
    $content .= '<h3>' .$post->title .'</h3>';
    $content .= $post->content;
}

echo $content;
```

For rendering the **Pagination Links**, the **Pagination\Paginator** use the concept of **Presenter**, by default using the native one, **Pagination\Presenter**, compatible with the **Bootstrap 3.x** based templates, but it is possible to extend it and to use a custom Presenter instance, as following:
```php
use App\Pagination\CustomPresenter;
use DB;

// Get a Paginator instance from the QueryBuilder.
$paginate = DB::table('posts')->paginate(20);

// Get a custom Presenter instance and setup it.
$presenter = new CustomPresenter($paginate);

echo $presenter->links();
```
Alternatively, you can setup the custom Presenter instance to be used by the Paginator instance directly by default. 
```php
$paginate->setPresenter($presenter);

// Print the custom Pagination Links.
echo $paginate->links();
```

To note that this **Pagination API** is a new feature of **New Style APIs** only, then it doesn't introduce any API break.

### CODE REVIEW

**Comment by jimgwhit** [comment]
@daveismyname and @LuckyCyborg why do we need another paginator, this style pagination is ALREADY there look at the page_eloq.md file. Already works perfect with eloquent and illuminate style.  
Is there no end to the madness?
```
namespace App\Controllers;

use Core\Controller;
use Core\View;
use Helpers\Url;
use Helpers\Session;
use Helpers\Csrf;
use Helpers\Paginator;

class Pet extends Controller
{
    private $pet;

    public function __construct()
    {
        parent::__construct();
        $this->pet = new \Models\PetModel();

        if (Session::get('loggin') == false) {
            Url::redirect('admin/login');
        }
    }

    public function index()
    {
        $petSearch = (isset($_REQUEST['psch']) != '' ? $_REQUEST['psch'] : "");
        Session::set('petSearch', $petSearch);

        $petRows = $this->pet->petCount($petSearch); // get number of rows
        Session::set('petRows', $petRows);

        $pages = new Paginator('5', 'p');
        $pages->setTotal($petRows);
        Session::set('petPage', $pages->getInstance());

        $data['pageLinks'] = $pages->pageLinks('?', '&psch=' . $petSearch);
        $data['title'] = 'Pet';
        $data['pets'] = $this->pet->getPets($pages->getLimit2(), $pages->getPerPage(), $petSearch); // get the data

        $this->view->renderTemplate('header', $data);
        $this->view->render('pet/index', $data);
        $this->view->renderTemplate('footer', $data);
    }

And typical examples of getting count and retrieving records in the model are:

public function petCount($petsearch = "")
{
    $petsearch = $petSearch . "%";

    return Capsule::table('pets')->where('petName', 'like', $petsearch)->count();
}

public function getPets($offset = "", $rowsPerPage = "", $petSearch = "")
{
    $petsearch = $petSearch . "%";

    return Capsule::table('pets')
                    ->where('petName', 'like', $petsearch)
                    ->orderBy('petName', 'asc')
                    ->skip($offset)->take($rowsPerPage)->get();
}
```

**Comment by DonWagner** [comment]
@jimgwhit -> it's getting old and pretty annoying to read all the time.
Don't use it, if you don't like it, no one asking you to use it.

File: app/Language/Ja/messages.php [hunk 1]
```diff
@@ -1,4 +1,5 @@
 <?php
 
 return array (
+  'Pagination' => '',
```
Comment (Kingorgg) [inline app/Language/Ja/messages.php]: Ja: 'pagination' => 'ページネーション',

**Comment by jimgwhit** [comment]
Well @LuckyCyborg you just fu---- me up the a-- again I don't see a checkbox on godaddy hosting for OpenSSL. Now I can no longer use the framework for the non--profit. I hope you are so so so so so so so so so so so so so so so so so so so so so happy. No telling how many other now can't use the framework on shared hosting.

**Comment by LuckyCyborg** [comment]
@jimgwhit First of all, this Paginator is exclusively for the New Style APIs, why the Hell Gods you comment there when you do not use those APIs?

Secondly, the Helpers\Paginator is hard-coded for Bootstrap 3, this one use a Presenter, then could be easily extended for another CSS toolkits.

Thirdly, I'm a very lazy person, I do not like to write supplementary code, just in the name of simplicity (!?!), when I can make the Framework to do the task itself. 

And in the case of **Pagination\Paginator**, the native QueryBuilder returns a ready to use Paginator instance. To replicate you code, for me is enough to write:

```php
$petsearch = $petSearch . "%";

$pets = DB::table('pets')
    ->where('petName', 'like', $petsearch)
    ->orderBy('petName', 'asc')
    ->paginate(25);

return View::make('Pet/Index')
    ->shares('title', 'Pet')
    ->withPets($pets);
```
Voila!

**Comment by LuckyCyborg** [comment]
@jimgwhit 

> Well @LuckyCyborg you just fu---- me up the a-- again

Nope, the community fu---- you, they wanted OpenSSL, me I wanted MCrypt.

Take care! :smirk:

**Comment by Kingorgg** [comment]
@jimgwhit Can't you just contact your host to enable it? Most hosts with good customer service will do that, you know.

**Comment by jimgwhit** [comment]
WELL @LuckyCyborg I AM BACK, Godaddy already has openssl implemented on their servers, so hi, hi, hi.  
Since I am back are you going to give a full model vew controller example of how I will have to implement your new paginator?

**Comment by DonWagner** [comment]
With the language you're not going to get much sucess. And what has he done to you, but to come up with a solution to your problem. Pull yourself together.

**Comment by DonWagner** [comment]
@jimgwhit nice... what are you, 5 years old? With the language you're not going to get much success. And what has @Kingorgg done to you? Nothing but to come up with a idea or a solution to your problem. Pull yourself together.

**Comment by jimgwhit** [comment]
@DonWagner quit wanting everything easy and built in, freaking roll up your sleeves and learn to code.  
I have been doing this stuff since the late 80's, started out in dbase 3.  
I actually get sick when people want everything built in.  Gee's that had to put in a auth system because people couldn't go to owasp site and actually read and learn php security techniques.

**Comment by Kingorgg** [comment]
@jimgwhit 
> Kingorgg if lucky f---- are you there to s---- it. It sounds as though you would agree with anything in the whole world if lucky liked it, right or wrong. Stop that, be your own person.

Where does it say in my comment that I agree or disagree with this PR? All I said was you can probably contact your host to get OpenSSL enabled. Learn to read....

**Comment by DonWagner** [comment]
If I could and had time, I would. But the fact is that I just have to code for small projects, therefore, is a framework that makes things easy it best for me. If i was a pro i would build my own framework, and not spend my time to complain about that someone will do something everyone can enjoy.

If you are so damn good, and do not like all the things that are in this framework, why do not you make your own framework? Obviously you think that you are damn good and us there is less good can not be allowed to make our own small sites from a framework.

The meaning of a framework is to help people to make sites easier and help others.

**Comment by jimgwhit** [comment]
@DonWagner If someone would get this thing stable and leave it alone for a while.  Why isn't there a stable branch and a seperate dev branch like laravel has. How can I actually get (with confidence) a site programmed with 20 PR's a week, and with 95% of the PR's being for things that people are too lazy to implement for themselves.  That Auth thing is still crawling under my skin.  But at least it's optional.  
But now all of a sudden he is coming out with breaking api's.  I just want a stable version that I can actually use without having to update it with 20 plus PR's weekly.

**Comment by jimgwhit** [comment]
> and do not like all the things that are in this framework  

I like the framework, but the original premise was KEEP IT SIMPLE. It is a shell for mvc.

**Comment by jimgwhit** [comment]
@Kingorgg they are right, you have been good, just upset at another PR.  Have you disagreed with any PR?

**Comment by LuckyCyborg** [comment]
@jimgwhit 

> I like the framework, but the original premise was KEEP IT SIMPLE. It is a shell for mvc.

**You don't like** this Nova Framework, Jim.

You are in denial, believing that Nova Framework is **a shell** for MVC. It is not **a shell** but a complete Framework with strong Routing, Middlewares, Hooks, Events, Cookies encryption, even a native QueryBuilder and another hundred things.

If you really need **a shell for MVC**, please do a favor yourself and to us, and look for another Framework, better appropriate to your needs or stuck on **SMVC v2.2**.

Look what you do there, for example, insulting the members of this community for nothing.

**Comment by DonWagner** [comment]
How can anyone be so focused on themselves and be so indifferent to everyone else ??
If the old version was so good, why don't you use that version and so ignore what else is done to the new version.?????
No one forces you to use what is made in the new version??

It's going to be stable. And yes much has happened, but only for good !! I would also like to spend my time on other things on my job and my family and daughter, so if I can save time on code then it could be great. I'm not in it to earn money it's a hobby to me.

I do not want to discuss with you anymore. It is sad and simplistic, and you can not talk properly to people, you never has anything positive to say and you can not see further than your own nose.
You can, at the same time, not understand that you are not forced to use the new stuff. You only have interest in telling how bad everything new is, despite the fact that you STILL are free to use the old methods or the old version of SMVC / NOVA.

So how do we ever get to hold a sensible conversation with you when you don't understand that, I do not know, but I do not bother anymore.

**Comment by LuckyCyborg** [comment]
@Kingorgg BTW, Daniel. I updated the Japanese string. Thanks!

**Comment by LuckyCyborg** [comment]
@DonWagner Thanks for your opinion, and I believe that Nova community is persons like you, like Daniel, like others and, first of all, like Dave. 

That **person** wandering there insulting and lamenting about every thingie, **he** is just just a unfortunate intruder in this wonderful community looking for things which do not exists.

**Comment by geomorillo** [comment]
Please close this since fighting is nonconstructive

**Comment by jimgwhit** [comment]
Look at PR #862 where I do give good replies when something makes good sense.  
I still say a new paginator is not needed all can be done in the current paginator.  
For crying out loud read the current documentation that's inside the current paGina tor I mean for crying out loud. In other words I'd gladly agree with something that makes good sense whereas this doesn't make any sense whatsoever.

**Comment by woodsy1** [comment]
Jim it might not make sense to you but it does to others.
It is made to work seamlessly with the new APIs and people want that.

**Comment by daveismyname** [comment]
Jim you've expressed your opinion a trillion times no one agrees you with, I cannot understand your viewpoint you bitch and moan that x feature makes no sense to have then demand help in using it. 

If you have nothing constructive to add then keep your negative thoughts to yourself. We all give up our own time to help build this together. it's not big or clever to belittle other users.

Lets leave it at that.

**Pull request merged**